src/hotspot/share/opto/vector.cpp

```diff
@@ -265,39 +264,54 @@ Node* PhaseVector::expand_vbox_node_helper(VectorBoxNode* vec_box,
     // move up and are guaranteed to dominate.
     Node* new_phi = new PhiNode(vbox->as_Phi()->region(), box_type);
     for (uint i = 1; i < vbox->req(); i++) {
-      Node* new_box = expand_vbox_node_helper(vec_box, vbox->in(i), vect, box_type, vect_type);
+      Node* new_box = expand_vbox_node_helper(vbox->in(i), vect, box_type, vect_type);
       new_phi->set_req(i, new_box);
     }
     new_phi = C->initial_gvn()->transform(new_phi);
     return new_phi;
   } else if (vbox->is_Proj() && vbox->in(0)->Opcode() == Op_VectorBoxAllocate) {
     VectorBoxAllocateNode* vbox_alloc = static_cast<VectorBoxAllocateNode*>(vbox->in(0));
-    return expand_vbox_alloc_node(vec_box, vbox_alloc, box_type, vect_type);
+    return expand_vbox_alloc_node(vbox_alloc, vect, box_type, vect_type);
   } else {
     assert(!vbox->is_Phi(), "");
     // TODO: assert that expanded vbox is initialized with the same value (vect).
     return vbox; // already expanded
   }
 }
 
-Node* PhaseVector::expand_vbox_alloc_node(VectorBoxNode* vec_box,
-                                          VectorBoxAllocateNode* vbox_alloc,
+Node* PhaseVector::expand_vbox_alloc_node(VectorBoxAllocateNode* vbox_alloc,
+                                          Node* vect,
                                           const TypeInstPtr* box_type,
                                           const TypeVect* vect_type) {
   JVMState* jvms = clone_jvms(C, vbox_alloc);
   GraphKit kit(jvms);
+  PhaseGVN& gvn = kit.gvn();
 
   ciInlineKlass* vk = static_cast<ciInlineKlass*>(box_type->inline_klass());
-  const TypeKlassPtr* klass_type = box_type->as_klass_type();
-  Node* klass_node = kit.makecon(klass_type);
-  Node* buffer_mem = kit.new_instance(klass_node, NULL, NULL, /* deoptimize_on_exception */ true);
-  vec_box->store(&kit, buffer_mem, buffer_mem, vk);
+
+  // Re-generate an InlineTypeNode to represent the payload field. This is necessary
+  // in case the input "vect" is not the original vector value when creating the
+  // VectorBox (e.g. original vector value is a PhiNode).
+  ciInlineKlass* payload = vk->declared_nonstatic_field_at(0)->type()->as_inline_klass();
+  Node* payload_value = InlineTypeNode::make_uninitialized(gvn, payload, true);
+  payload_value->as_InlineType()->set_field_value(0, vect);
+  payload_value = gvn.transform(payload_value);
+
+  // Re-generate an InlineTypeNode to represent the vector object. New a buffer
+  // and save its field value to the buffer.
+  InlineTypeNode* vector = InlineTypeNode::make_uninitialized(gvn, vk, false);
+  vector->set_field_value(0, payload_value);
+  vector = gvn.transform(vector)->as_InlineType();
+
+  Node* klass_node = kit.makecon(TypeKlassPtr::make(vk));
+  Node* alloc_oop  = kit.new_instance(klass_node, NULL, NULL, /* deoptimize_on_exception */ true, vector);
+  vector->store(&kit, alloc_oop, alloc_oop, vk);
```

Comment:
Do you think we should allocate explicit payload for non-flattened case before calling InlineTypeNode::store.
As per following 
https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/classfile/fieldLayoutBuilder.cpp#L768
Flattening depends on InlineFieldMaxFlatSize (default 128 bytes so fine for current vectors) but its configurable and value can be reduced.

Reply:
Yes, allocating a buffer is needed. And the `InlineTypeNode::store()` can handled such cases. If the field type is an inline type and is not flattened, it goes to https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/opto/inlinetypenode.cpp#L518, which will create a buffer for the InlineTypeNode. Please see: https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/opto/graphKit.cpp#L1656

I also tested the case with `-XX:InlineFieldMaxFlatSize=0` to force it not be flattened. The VectorBox expanding code is the expected one like before, which includes both the payload buffer and vector buffer and the relative stores. 

BTW, currently this flag cannot work well since there is issue when handling this option. You may need to apply the following changes when testing:
```
diff --git a/src/hotspot/share/classfile/fieldLayoutBuilder.cpp b/src/hotspot/share/classfile/fieldLayoutBuilder.cpp
index 516ad9bb7..538f31ae6 100644
--- a/src/hotspot/share/classfile/fieldLayoutBuilder.cpp
+++ b/src/hotspot/share/classfile/fieldLayoutBuilder.cpp
@@ -774,7 +774,7 @@ void FieldLayoutBuilder::regular_field_sorting() {
               //too_volatile_to_flatten = false; //FIXME
               // volatile fields are currently never inlined, this could change in the future
             }
-            if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) || fs.access_flags().is_final()) {
+            if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) && fs.access_flags().is_final()) {
               group->add_inlined_field(fs, vk);
               _nonstatic_oopmap_count += vk->nonstatic_oop_map_count();
               fs.set_inlined(true);
@@ -877,7 +877,7 @@ void FieldLayoutBuilder::inline_class_field_sorting(TRAPS) {
               //too_volatile_to_flatten = false; //FIXME
               // volatile fields are currently never inlined, this could change in the future
             }
-            if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) || fs.access_flags().is_final()) {
+            if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) && fs.access_flags().is_final()) {
               group->add_inlined_field(fs, vk);
               _nonstatic_oopmap_count += vk->nonstatic_oop_map_count();
               field_alignment = vk->get_alignment();
```

Also there are other issues when the payload is not flattened. But it seems not related to the box/unbox expanding part. I'v recorded the issues and plan to look at in future. You can also take a look if interested. Thanks!

Reply:
Thanks for your explanations, yes for non-flattened case buffering happens underneath _access_store_at_